{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "from pysam import VariantFile\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "mydata = \"../ValidationData/mydata/original.snpeff.state.disease.identifiedgene.filtered.splai.tsv\"\n",
    "df = pd.read_table(mydata, sep='\\t', dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_bcf = \"../clinvar/Filtered_BCF_GRCh37_20241211-044124/clinvar_GRCh37.germline.nocoflicted.bcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1-100523565-101533565-C-T'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i = 2\n",
    "\n",
    "query_chr = df.iloc[i, 34]\n",
    "query_pos = df.iloc[i, 35]\n",
    "query_ref = df.iloc[i, 36]\n",
    "query_alt = df.iloc[i, 37]\n",
    "\n",
    "query_start = int(query_pos) - 10000\n",
    "query_end = int(query_pos) + 1000000\n",
    "query_variant = f\"{query_chr}-{query_start}-{query_end}-{query_ref}-{query_alt}\"\n",
    "\n",
    "query_variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcf_in = VariantFile(clinvar_bcf)\n",
    "recs = bcf_in.fetch(query_chr, query_start, query_end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1-100573865-T-A\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Benign']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "rec = next(recs)\n",
    "print(f\"{rec.contig}-{rec.pos}-{rec.ref}-{rec.alts[0]}\")\n",
    "clnsig = rec.info[\"CLNSIG\"]\n",
    "clnsig: tuple = ('Pathogenic', 'Pathogenic/LP', 'Benign')\n",
    "\n",
    "clnsigs = [x for x in rec.info[\"CLNSIG\"]]\n",
    "clnsigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "cln_bcf = VariantFile(clinvar_bcf)\n",
    "\n",
    "def remove_square_brackets(s: str) -> str:\n",
    "    return s.replace(\"[\", \"\").replace(\"]\", \"\")\n",
    "\n",
    "def anno_same_pos_vars(row, cln_bcf: VariantFile) -> str:\n",
    "    samepos = []\n",
    "    query_chr: str = f\"{row['CHROM']}\"\n",
    "    query_pos: int = int(row['POS'])\n",
    "    query_ref: str = row['REF']\n",
    "    query_alt: str = row['ALT']\n",
    "    query_variant: str = f\"{query_chr}-{query_pos}-{query_ref}-{query_alt}\"\n",
    "    query_start: int = query_pos - 1\n",
    "    query_end: int = query_pos\n",
    "    recs = cln_bcf.fetch(query_chr, query_start, query_end)\n",
    "\n",
    "    # main loop \n",
    "    while 1:\n",
    "        try:\n",
    "            rec = next(recs)\n",
    "        except StopIteration:\n",
    "            break\n",
    "        else:\n",
    "            if rec.alts is None:\n",
    "                rec_alt: str = \".\"\n",
    "            else:\n",
    "                rec_alt: str = rec.alts[0]\n",
    "            rec_id = f\"{rec.contig}-{rec.pos}-{rec.ref}-{rec_alt}\"\n",
    "            if query_variant == rec_id:\n",
    "                clnsigs: list = [x for x in rec.info[\"CLNSIG\"]]\n",
    "                samepos.append(clnsigs)\n",
    "            else:\n",
    "                pass\n",
    "    \n",
    "    # Return the results\n",
    "    if samepos == []:\n",
    "        return \"No_ClinVar_info_found\"\n",
    "    else:\n",
    "        return remove_square_brackets(str(samepos))\n",
    "\n",
    "df['same_pos'] = df.apply(anno_same_pos_vars, cln_bcf=cln_bcf, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _generate_query_pos(row) -> tuple:\n",
    "    if row['SpliceType'] == 'Donor_int':\n",
    "        if row['Strand'] == '+':\n",
    "            query_start: int = query_pos - int(row['IntronDist']) - 2 - 1\n",
    "            query_end: int = query_pos - int(row['IntronDist']) + 6\n",
    "        elif row['Strand'] == '-':\n",
    "            query_start: int = query_pos + int(row['IntronDist']) - 6 - 1\n",
    "            query_end: int = query_pos + int(row['IntronDist']) + 2\n",
    "        else:\n",
    "            return 'unk_Strand', 'unk_Strand', 'unk_Strand'\n",
    "\n",
    "    elif row['SpliceType'] == 'Donor_ex':\n",
    "        if row['Strand'] == '+':\n",
    "            query_start: int = query_pos + int(row['exon_pos']) - 3 - 1\n",
    "            query_end: int = query_pos + int(row['exon_pos']) + 5\n",
    "        elif row['Strand'] == '-':\n",
    "            query_start: int = query_pos - int(row['exon_pos']) - 5 - 1\n",
    "            query_end: int = query_pos - int(row['exon_pos']) + 3\n",
    "        else:\n",
    "            return 'unk_Strand', 'unk_Strand', 'unk_Strand'\n",
    "    \n",
    "    elif row['SpliceType'] == 'Acceptor_int':\n",
    "        if row['Strand'] == '+':\n",
    "            query_start: int = query_pos + (- int(row['IntronDist'])) - 20 - 1\n",
    "            query_end: int = query_pos + (- int(row['IntronDist'])) + 0\n",
    "        elif row['Strand'] == '-':\n",
    "            query_start: int = query_pos - (- int(row['IntronDist'])) - 0 - 1\n",
    "            query_end: int = query_pos - (- int(row['IntronDist'])) + 20\n",
    "        else:\n",
    "            return 'unk_Strand', 'unk_Strand', 'unk_Strand'\n",
    "    \n",
    "    elif row['SpliceType'] == 'Acceptor_ex':\n",
    "        if row['Strand'] == '+':\n",
    "            query_start: int = query_pos - int(row['exon_pos']) - 19 - 1\n",
    "            query_end: int = query_pos - int(row['exon_pos']) + 1\n",
    "        elif row['Strand'] == '-':\n",
    "            query_start: int = query_pos + int(row['exon_pos']) - 1 - 1\n",
    "            query_end: int = query_pos + int(row['exon_pos']) + 19\n",
    "        else:\n",
    "            return 'unk_Strand'\n",
    "        \n",
    "    else:\n",
    "        return 'unk_SpliceType', 'unk_SpliceType', 'unk_SpliceType'\n",
    "    \n",
    "    return str(row['CHROM']), query_start, query_end\n",
    "\n",
    "def anno_same_motif_vars(row, cln_bcf: VariantFile) -> str:\n",
    "    region: tuple = _generate_query_pos(row)\n",
    "\n",
    "    if region[0] == 'unk_Strand':\n",
    "        return 'unk_Strand'\n",
    "    elif region[0] == 'unk_SpliceType':\n",
    "        return 'unk_SpliceType'\n",
    "    else:\n",
    "        recs = cln_bcf.fetch(*region)\n",
    "\n",
    "    # Main loop\n",
    "    samemotifs = []\n",
    "    while 1:\n",
    "        try:\n",
    "            clinvar = next(recs)\n",
    "        except StopIteration:\n",
    "            break\n",
    "        else:\n",
    "            registered_var = f'{clinvar[7]}_{clinvar[0]}:{clinvar[1]}'\n",
    "            samemotifs.append(registered_var)\n",
    "\n",
    "    if samemotifs == []:\n",
    "        return \"No_ClinVar_info_found\"\n",
    "    else:\n",
    "        return remove_square_brackets(str(samemotifs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clinvars = cln_bcf.fetch(\n",
    "#          query_chr, query_start, query_end, parser=pysam.asBed())\n",
    "i = 3\n",
    "regions = \"2\", 100, 2000000\n",
    "\n",
    "recs = cln_bcf.fetch(*regions)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def anno_same_motif_vars(row, tabixfile: pysam.pysam.libctabix.TabixFile):\n",
    "    \"\"\" Fetch the variants within the same splicing motif from ClinVar data\n",
    "\n",
    "    Args:\n",
    "        row (_type_): _description_\n",
    "\n",
    "    Returns:\n",
    "        list: Return a list of variants registered in ClinVar \n",
    "                within the same splicing motif (Prediction_CHROM:POS)\n",
    "    \"\"\"\n",
    "    \n",
    "    # Initialize result list & set common variables\n",
    "    samemotifs = []\n",
    "    query_chr: str = f'{row[\"CHROM\"]}'\n",
    "    query_pos: int = int(row['POS'])\n",
    "\n",
    "    # Generate query positions \n",
    "    query_chr, query_start, query_end = _generate_query_pos(row)\n",
    "    if row['SpliceType'] == 'Donor_int':\n",
    "        if row['Strand'] == '+':\n",
    "            query_start: int = query_pos - int(row['IntronDist']) - 2 - 1\n",
    "            query_end: int = query_pos - int(row['IntronDist']) + 6\n",
    "        elif row['Strand'] == '-':\n",
    "            query_start: int = query_pos + int(row['IntronDist']) - 6 - 1\n",
    "            query_end: int = query_pos + int(row['IntronDist']) + 2\n",
    "        else:\n",
    "            return 'unk_Strand', 'unk_Strand', 'unk_Strand'\n",
    "\n",
    "    elif row['SpliceType'] == 'Donor_ex':\n",
    "        if row['Strand'] == '+':\n",
    "            query_start: int = query_pos + int(row['exon_pos']) - 3 - 1\n",
    "            query_end: int = query_pos + int(row['exon_pos']) + 5\n",
    "        elif row['Strand'] == '-':\n",
    "            query_start: int = query_pos - int(row['exon_pos']) - 5 - 1\n",
    "            query_end: int = query_pos - int(row['exon_pos']) + 3\n",
    "        else:\n",
    "            return 'unk_Strand', 'unk_Strand', 'unk_Strand'\n",
    "    \n",
    "    elif row['SpliceType'] == 'Acceptor_int':\n",
    "        if row['Strand'] == '+':\n",
    "            query_start: int = query_pos + (- int(row['IntronDist'])) - 20 - 1\n",
    "            query_end: int = query_pos + (- int(row['IntronDist'])) + 0\n",
    "        elif row['Strand'] == '-':\n",
    "            query_start: int = query_pos - (- int(row['IntronDist'])) - 0 - 1\n",
    "            query_end: int = query_pos - (- int(row['IntronDist'])) + 20\n",
    "        else:\n",
    "            return 'unk_Strand', 'unk_Strand', 'unk_Strand'\n",
    "    \n",
    "    elif row['SpliceType'] == 'Acceptor_ex':\n",
    "        if row['Strand'] == '+':\n",
    "            query_start: int = query_pos - int(row['exon_pos']) - 19 - 1\n",
    "            query_end: int = query_pos - int(row['exon_pos']) + 1\n",
    "        elif row['Strand'] == '-':\n",
    "            query_start: int = query_pos + int(row['exon_pos']) - 1 - 1\n",
    "            query_end: int = query_pos + int(row['exon_pos']) + 19\n",
    "        else:\n",
    "            return 'unk_Strand', 'unk_Strand', 'unk_Strand'\n",
    "        \n",
    "    else:\n",
    "        return 'unk_SpliceType', 'unk_SpliceType', 'unk_SpliceType'\n",
    "\n",
    "    clinvars = tabixfile.fetch(\n",
    "        query_chr, query_start, query_end, parser=pysam.asBed())\n",
    "        \n",
    "    # Main loop\n",
    "    while 1:\n",
    "        try:\n",
    "            clinvar = next(clinvars)\n",
    "        except StopIteration:\n",
    "            break\n",
    "        else:\n",
    "            registered_var = f'{clinvar[7]}_{clinvar[0]}:{clinvar[1]}'\n",
    "            samemotifs.append(registered_var)\n",
    "\n",
    "    return samemotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "same_pos\n",
       "No_ClinVar_info_found             15716\n",
       "'Likely_benign'                     539\n",
       "'Pathogenic'                         35\n",
       "'Likely_pathogenic'                  22\n",
       "'Pathogenic/Likely_pathogenic'       20\n",
       "'Benign'                             14\n",
       "'Benign/Likely_benign'                7\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['same_pos'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>fa</th>\n",
       "      <th>mo</th>\n",
       "      <th>type</th>\n",
       "      <th>ID_x</th>\n",
       "      <th>vqslod</th>\n",
       "      <th>triodenovo</th>\n",
       "      <th>dnmfilter</th>\n",
       "      <th>denovogear</th>\n",
       "      <th>denovofilter</th>\n",
       "      <th>...</th>\n",
       "      <th>DS_AG</th>\n",
       "      <th>DS_AL</th>\n",
       "      <th>DS_DG</th>\n",
       "      <th>DS_DL</th>\n",
       "      <th>DP_AG</th>\n",
       "      <th>DP_AL</th>\n",
       "      <th>DP_DG</th>\n",
       "      <th>DP_DL</th>\n",
       "      <th>maxsplai</th>\n",
       "      <th>same_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Sample_10564</td>\n",
       "      <td>Sample_10562</td>\n",
       "      <td>Sample_10563</td>\n",
       "      <td>snv</td>\n",
       "      <td>1:100661991-A-T</td>\n",
       "      <td>-1.171</td>\n",
       "      <td>2.19</td>\n",
       "      <td>0.0235918193666265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-2</td>\n",
       "      <td>-13</td>\n",
       "      <td>-13</td>\n",
       "      <td>-221</td>\n",
       "      <td>0.04</td>\n",
       "      <td>[[Likely_benign]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Sample_10564</td>\n",
       "      <td>Sample_10562</td>\n",
       "      <td>Sample_10563</td>\n",
       "      <td>snv</td>\n",
       "      <td>1:100661994-A-T</td>\n",
       "      <td>-0.6796</td>\n",
       "      <td>2.19</td>\n",
       "      <td>0.00131575948972965</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-37</td>\n",
       "      <td>-16</td>\n",
       "      <td>-16</td>\n",
       "      <td>-224</td>\n",
       "      <td>0.04</td>\n",
       "      <td>[[Likely_benign]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Sample_10564</td>\n",
       "      <td>Sample_10562</td>\n",
       "      <td>Sample_10563</td>\n",
       "      <td>snv</td>\n",
       "      <td>1:100661997-A-G</td>\n",
       "      <td>-1.255</td>\n",
       "      <td>10.78</td>\n",
       "      <td>0.000628006845710701</td>\n",
       "      <td>0.998948</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-19</td>\n",
       "      <td>92</td>\n",
       "      <td>-227</td>\n",
       "      <td>4</td>\n",
       "      <td>0.06</td>\n",
       "      <td>[[Likely_benign]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Sample_7118</td>\n",
       "      <td>Sample_7119</td>\n",
       "      <td>Sample_7120</td>\n",
       "      <td>snv</td>\n",
       "      <td>1:109477385-G-A</td>\n",
       "      <td>3.68</td>\n",
       "      <td>13.23</td>\n",
       "      <td>0.998983274802696</td>\n",
       "      <td>0.999996</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1638</td>\n",
       "      <td>152</td>\n",
       "      <td>-143</td>\n",
       "      <td>-196</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[[Likely_benign]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Sample_7118</td>\n",
       "      <td>Sample_7119</td>\n",
       "      <td>Sample_7120</td>\n",
       "      <td>snv</td>\n",
       "      <td>1:109477385-G-A</td>\n",
       "      <td>3.68</td>\n",
       "      <td>13.23</td>\n",
       "      <td>0.998983274802696</td>\n",
       "      <td>0.999996</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1638</td>\n",
       "      <td>152</td>\n",
       "      <td>-143</td>\n",
       "      <td>-196</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[[Likely_benign]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16313</th>\n",
       "      <td>Sample_17213</td>\n",
       "      <td>Sample_17214</td>\n",
       "      <td>Sample_17215</td>\n",
       "      <td>snv</td>\n",
       "      <td>X:70612508-C-T</td>\n",
       "      <td>-1.029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>703</td>\n",
       "      <td>189</td>\n",
       "      <td>111</td>\n",
       "      <td>-38</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[[Likely_benign]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16322</th>\n",
       "      <td>Sample_12995</td>\n",
       "      <td>Sample_12996</td>\n",
       "      <td>Sample_12997</td>\n",
       "      <td>snv</td>\n",
       "      <td>X:71825181-G-A</td>\n",
       "      <td>0.395</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.49</td>\n",
       "      <td>3049</td>\n",
       "      <td>69</td>\n",
       "      <td>-2935</td>\n",
       "      <td>-60</td>\n",
       "      <td>0.49</td>\n",
       "      <td>[[Pathogenic]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16340</th>\n",
       "      <td>Sample_12362</td>\n",
       "      <td>Sample_12363</td>\n",
       "      <td>Sample_12364</td>\n",
       "      <td>snv</td>\n",
       "      <td>X:85218829-A-G</td>\n",
       "      <td>1.17</td>\n",
       "      <td>15.13</td>\n",
       "      <td>0.189823675562113</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>148</td>\n",
       "      <td>-1853</td>\n",
       "      <td>-159</td>\n",
       "      <td>-1989</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[[Benign]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16341</th>\n",
       "      <td>Sample_5766</td>\n",
       "      <td>Sample_5784</td>\n",
       "      <td>Sample_5785</td>\n",
       "      <td>snv</td>\n",
       "      <td>X:85282539-T-G</td>\n",
       "      <td>-0.7864</td>\n",
       "      <td>14.82</td>\n",
       "      <td>0.506066077390227</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22</td>\n",
       "      <td>-2</td>\n",
       "      <td>-125</td>\n",
       "      <td>2260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[[Likely_benign]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16349</th>\n",
       "      <td>Sample_20591</td>\n",
       "      <td>Sample_20592</td>\n",
       "      <td>Sample_20593</td>\n",
       "      <td>snv</td>\n",
       "      <td>X:9714186-G-A</td>\n",
       "      <td>-0.2732</td>\n",
       "      <td>15.43</td>\n",
       "      <td>0.839352299143052</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-199</td>\n",
       "      <td>-514</td>\n",
       "      <td>-2581</td>\n",
       "      <td>402</td>\n",
       "      <td>0.01</td>\n",
       "      <td>[[Likely_benign]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>637 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample            fa            mo type             ID_x  \\\n",
       "6      Sample_10564  Sample_10562  Sample_10563  snv  1:100661991-A-T   \n",
       "7      Sample_10564  Sample_10562  Sample_10563  snv  1:100661994-A-T   \n",
       "8      Sample_10564  Sample_10562  Sample_10563  snv  1:100661997-A-G   \n",
       "35      Sample_7118   Sample_7119   Sample_7120  snv  1:109477385-G-A   \n",
       "36      Sample_7118   Sample_7119   Sample_7120  snv  1:109477385-G-A   \n",
       "...             ...           ...           ...  ...              ...   \n",
       "16313  Sample_17213  Sample_17214  Sample_17215  snv   X:70612508-C-T   \n",
       "16322  Sample_12995  Sample_12996  Sample_12997  snv   X:71825181-G-A   \n",
       "16340  Sample_12362  Sample_12363  Sample_12364  snv   X:85218829-A-G   \n",
       "16341   Sample_5766   Sample_5784   Sample_5785  snv   X:85282539-T-G   \n",
       "16349  Sample_20591  Sample_20592  Sample_20593  snv    X:9714186-G-A   \n",
       "\n",
       "        vqslod triodenovo             dnmfilter denovogear denovofilter  ...  \\\n",
       "6       -1.171       2.19    0.0235918193666265        NaN        False  ...   \n",
       "7      -0.6796       2.19   0.00131575948972965        NaN        False  ...   \n",
       "8       -1.255      10.78  0.000628006845710701   0.998948        False  ...   \n",
       "35        3.68      13.23     0.998983274802696   0.999996         True  ...   \n",
       "36        3.68      13.23     0.998983274802696   0.999996         True  ...   \n",
       "...        ...        ...                   ...        ...          ...  ...   \n",
       "16313   -1.029        NaN                   NaN        NaN         True  ...   \n",
       "16322    0.395        NaN                   NaN        NaN        False  ...   \n",
       "16340     1.17      15.13     0.189823675562113   0.999997         True  ...   \n",
       "16341  -0.7864      14.82     0.506066077390227        1.0         True  ...   \n",
       "16349  -0.2732      15.43     0.839352299143052   0.999976         True  ...   \n",
       "\n",
       "      DS_AG DS_AL DS_DG DS_DL DP_AG  DP_AL  DP_DG  DP_DL maxsplai  \\\n",
       "6       0.0  0.04   0.0  0.02    -2    -13    -13   -221     0.04   \n",
       "7       0.0  0.04   0.0  0.01   -37    -16    -16   -224     0.04   \n",
       "8      0.06   0.0  0.02   0.0   -19     92   -227      4     0.06   \n",
       "35      0.0   0.0   0.0   0.0  1638    152   -143   -196      0.0   \n",
       "36      0.0   0.0   0.0   0.0  1638    152   -143   -196      0.0   \n",
       "...     ...   ...   ...   ...   ...    ...    ...    ...      ...   \n",
       "16313   0.0   0.0   0.0   0.0   703    189    111    -38      0.0   \n",
       "16322   0.0  0.38   0.0  0.49  3049     69  -2935    -60     0.49   \n",
       "16340   0.0   0.0   0.0   0.0   148  -1853   -159  -1989      0.0   \n",
       "16341   0.0   0.0   0.0   0.0    22     -2   -125   2260      0.0   \n",
       "16349  0.01   0.0   0.0   0.0  -199   -514  -2581    402     0.01   \n",
       "\n",
       "                same_pos  \n",
       "6      [[Likely_benign]]  \n",
       "7      [[Likely_benign]]  \n",
       "8      [[Likely_benign]]  \n",
       "35     [[Likely_benign]]  \n",
       "36     [[Likely_benign]]  \n",
       "...                  ...  \n",
       "16313  [[Likely_benign]]  \n",
       "16322     [[Pathogenic]]  \n",
       "16340         [[Benign]]  \n",
       "16341  [[Likely_benign]]  \n",
       "16349  [[Likely_benign]]  \n",
       "\n",
       "[637 rows x 69 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['same_pos'] != \"No_ClinVar_info_found\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nar",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
